rust/psibase_macros/src/graphql_macro.rs

```diff
@@ -0,0 +1,195 @@
+    parse_macro_input,
+    punctuated::Punctuated,
+    AttrStyle, ImplItem, Item, Result, Token, TypeParam,
+};
+
+#[derive(Debug)]
+struct Args {
+    query_name: Ident,
+    table: Ident,
+    key_fn: TypeParam,
+    subindex_params: Vec<TypeParam>,
+}
+
+impl Parse for Args {
+    fn parse(input: ParseStream) -> Result<Self> {
+        let mut vars = Punctuated::<TypeParam, Token![,]>::parse_terminated(input)?.into_iter();
+
+        let query_name = vars.next().expect("Query name not present").ident;
+        let table = vars.next().expect("Table name not present").ident;
+        let key_fn = vars.next().expect("Table key function not present");
+        let subindex_params = vars.collect();
+
+        Ok(Args {
+            query_name,
+            table,
+            key_fn,
+            subindex_params,
+        })
+    }
+}
+
+pub fn queries_macro_impl(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    let item = parse_macro_input!(item as Item);
+
+    let mut queries = quote! {};
+
+    match item {
+        Item::Impl(mut query_impl) => {
+            for impl_item in query_impl.items.iter_mut() {
+                if let ImplItem::Macro(macro_item) = impl_item {
+                    let query_doc = macro_item
+                        .attrs
+                        .iter()
+                        .filter(|attr| {
+                            matches!(attr.style, AttrStyle::Outer) && attr.path.is_ident("doc")
+                        })
+                        .fold(quote! {}, |a, b| quote! {#a #b});
+
+                    let macro_args = macro_item.mac.tokens.clone();
+
+                    let macro_str = macro_item.mac.path.to_token_stream().to_string();
+                    let query_fn: proc_macro2::TokenStream = if macro_str == "table_query" {
+                        table_query_macro_impl(macro_args.into()).into()
+                    } else if macro_str == "table_query_subindex" {
+                        table_query_subindex_macro_impl(macro_args.into()).into()
+                    } else {
+                        abort!(impl_item, "invalid table_query macro; expected: (table_query | table_query_subindex)");
+                    };
+
+                    queries = quote! {
+                        #queries
+
+                        #query_doc
+                        #query_fn
+                    };
+                }
+            }
+        }
+        _ => {
+            abort!(item, "queries attribute may only be used on an Impl")
+        }
+    }
+
+    let impl_query = quote! {
+        struct Query;
+
+        #[async_graphql::Object]
+        impl Query {
+            #queries
+        }
+    };
+
+    impl_query.into()
+}
+
+pub fn table_query_macro_impl(item: TokenStream) -> TokenStream {
+    let args = parse_macro_input!(item as Args);
+
+    let query_name = args.query_name;
+    let table = args.table;
+    let record = Ident::new(format!("{}Record", table).as_str(), table.span());
```

Comment:
Requiring a specific naming convention seems dangerous. Can't it be deduced as an associated type of the table type?

Reply:
We don't necessarily need to insert the original name of the type, though. We only need a sequence of tokens that will resolve to that type.